crates/exex/exex/src/manager.rs

```diff
@@ -303,83 +312,125 @@ impl ExExManager {
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.get_mut();
+
+        // Handle incoming ExEx events
+        for exex in &mut this.exex_handles {
+            while let Poll::Ready(Some(event)) = exex.receiver.poll_recv(cx) {
+                debug!(exex_id = %exex.id, ?event, "Received event from ExEx");
+                exex.metrics.events_sent_total.increment(1);
+                match event {
+                    ExExEvent::FinishedHeight(height) => exex.finished_height = Some(height),
+                }
+            }
+        }
+
         // Drain the finalized header stream and grab the last finalized header
         let mut last_finalized_header = None;
-        while let Poll::Ready(finalized_header) = self.finalized_header_stream.poll_next_unpin(cx) {
+        while let Poll::Ready(finalized_header) = this.finalized_header_stream.poll_next_unpin(cx) {
             last_finalized_header = finalized_header;
         }
-        // If there is a finalized header, finalize the WAL with it
         if let Some(header) = last_finalized_header {
-            self.wal.finalize((header.number, header.hash()).into())?;
+            debug!(header = ?header.num_hash(), "Received finalized header");
+
+            // Check if all ExExes are on the canonical chain
+            let exex_finished_heights = this
+                .exex_handles
+                .iter()
+                // Get ExEx ID and hash of the finished height for each ExEx
+                .map(|exex_handle| {
+                    (&exex_handle.id, exex_handle.finished_height.map(|block| block.hash))
+                })
+                // Deduplicate all hashes
+                .unique_by(|(_, hash)| *hash)
+                // Check if hashes are canonical
+                .map(|(exex_id, hash)| {
+                    hash.map_or(Ok((exex_id, hash, false)), |hash| {
+                        this.provider
+                            .header(&hash)
+                            // Save the ExEx ID, hash of the finished height, and whether the hash
+                            // is canonical
+                            .map(|header| (exex_id, Some(hash), header.is_some()))
+                    })
+                })
+                .collect::<Result<Vec<_>, _>>()?;
+            if exex_finished_heights.iter().all(|(_, _, is_canonical)| *is_canonical) {
+                // If there is a finalized header and all ExExs are on the canonical chain, finalize
+                // the WAL with the new finalized header
+                this.wal.finalize(header.num_hash())?;
+            } else {
+                let unfinalized_exexes = exex_finished_heights
+                    .into_iter()
+                    .filter_map(|(exex_id, hash, is_canonical)| {
+                        is_canonical.not().then_some((exex_id, hash))
+                    })
+                    .collect::<Vec<_>>();
```

Comment:
this collect is redundant, itertools has helper for this

Reply:
the collect is for debug logging below https://github.com/paradigmxyz/reth/blob/03fcdb4cda4aee28ff9762cca28a7c2dad3c90f5/crates/exex/exex/src/manager.rs#L369-L378